tools/smithy-rs-sync/src/main.rs

```diff
@@ -0,0 +1,409 @@
+const COMMIT_HASH_FILENAME: &str = ".smithyrs-githash";
+
+/// A macro for attaching info to error messages pointing to the line of code responsible for the error
+/// [Thanks to dtolnay for this macro](https://github.com/dtolnay/anyhow/issues/22#issuecomment-542309452)
+macro_rules! here {
+    () => {
+        concat!("error at ", file!(), ":", line!(), ":", column!())
+    };
+}
+
+/// Run this app in order to keep aws-sdk-rust in sync with smithy-rs.
+///
+/// pre-requisites:
+/// - an up-to-date local copy of smithy-rs repo
+/// - an up-to-date local copy of aws-sdk-rs repo
+/// - a Unix-ey system (for the `cp` and `rf` commands to work)
+/// - Java Runtime Environment v11 (in order to run gradle commands)
+///
+/// ```sh
+/// cargo run -- \
+/// --smithy-rs /Users/zhessler/Documents/smithy-rs-test/ \
+/// --aws-sdk /Users/zhessler/Documents/aws-sdk-rust-test/
+/// ```
+fn main() {
+    let Opt {
+        smithy_rs,
+        aws_sdk,
+        branch,
+    } = Opt::from_args();
+
+    if let Err(e) = sync_aws_sdk_with_smithy_rs(&smithy_rs, &aws_sdk, &branch) {
+        eprintln!("Sync failed with error: {:?}", e);
+    };
+}
+
+/// Run through all commits made to `smithy-rs` since last sync and "replay" them onto `aws-sdk-rust`
+fn sync_aws_sdk_with_smithy_rs(smithy_rs: &Path, aws_sdk: &Path, branch: &str) -> Result<()> {
+    // Open the repositories we'll be working with
+    let smithy_rs_repo = Repository::open(smithy_rs).context("couldn't open smithy-rs repo")?;
+    let aws_sdk_repo = Repository::open(aws_sdk).context("couldn't open aws-sdk-rust repo")?;
+
+    // Check repo that we're going to be moving the code into to see what commit it was last synced with
+    let last_synced_commit =
+        get_last_synced_commit(aws_sdk).context("couldn't get last synced commit")?;
+    let commit_revs = commits_to_be_applied(&smithy_rs_repo, &last_synced_commit)
+        .context("couldn't build list of commits that need to be synced")?;
+
+    if commit_revs.is_empty() {
+        println!("There are no new commits to be applied, have a nice day.");
+        return Ok(());
+    }
+
+    // `git checkout` the branch of `aws-sdk-rust` that we want to replay commits onto.
+    // By default, this is the `next` branch.
+    checkout_branch_to_sync_to(&aws_sdk_repo, branch).with_context(|| {
+        format!(
+            "couldn't checkout {} branch of aws-sdk-rust for syncing",
+            branch,
+        )
+    })?;
+
+    let number_of_commits = commit_revs.len();
+    println!(
+        "Found {} unsynced commit(s), syncing now...",
+        number_of_commits
+    );
+    // Run through all the new commits, syncing them one by one
+    for (i, rev) in commit_revs.iter().enumerate() {
+        let commit = smithy_rs_repo
+            .find_commit(*rev)
+            .with_context(|| format!("couldn't find commit {} in smithy-rs", rev))?;
+
+        println!("[{}/{}]\tsyncing {}...", i + 1, number_of_commits, rev);
```

Comment:
I'd probably use `eprintln` or tracing everywhere in case we decide to use stdout for structured output